<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2024-1</h1>

<h2>Alumno: Rivas Moreno Javier </h2>

<h1 align="center">Programación para la física computacional</h1>

<h2> Profesor: Sergio A. Alcalá Corona </h2>

---
### Ayudantes:
#### Daniel Rojo Mata
#### Rodrigo Vega Vilchis
---

## Práctica 1.  Programación para la física computacional


### Programación básica para la física

---

### EJERCICIOS:

**1. Caida de una pelota desde una torre:** Se deja caer una pelota desde una torre de altura $h$. Tiene velocidad inicial cero y acelera hacia abajo con la gravedad.

Escribe un programa que le pida al usuario que ingrese la altura en metros de la torre y luego calcule e imprima el tiempo en segundos hasta que la pelota toque el suelo (ignora la resistencia del aire).
    
Después, usa tu programa para calcular el tiempo de una pelota lanzada desde una torre de $100m$ de altura.

In [1]:
# Programa que pida una altura en metros de una torre e imprima el tiepo en segundos hasta que una pelota toque el suelo
import math

# Pedimos el valor de la altura
h = float(input("Ingrese la altura de la torre en metros:"))

# Realizamos el cálculo tomando en cuenta la gravedad de la Tierra
g = 9.81
t = math.sqrt((2*h)/g)

print("Dado un edificio de ", h, "metros de altura, al dejar caer una pelota el tiempo en llegar al suelo será de t = ", t, "segundos.")

Ingrese la altura de la torre en metros:100
Dado un edificio de  100.0 metros de altura, al dejar caer una pelota el tiempo en llegar al suelo será de t =  4.515236409857309 segundos.


**2. Altitud de un satélite**: se va a lanzar un satélite en una órbita circular alrededor de la Tierra de modo que orbite el planeta una vez cada $T$ segundos.

(a) Demuestre que la altitud $h$ sobre la superficie de la Tierra que debe tener el satélite es:

<h3 align="center">$h = \biggl( {GMT^2\over4\pi^2} \biggr)^{1/3} - R$,</h3>

donde $G=6.67\times10^{-11}\,\textrm{m}^3\,\textrm{kg}^{-1}\,\textrm{s}^{-2}$ es la constante gravitatoria de Newton, $M=5.97\times10^{24}\,$kg es la masa de la Tierra y $R=6371\,$km  es su radio.



**Demostración**

Partimos de le ecuación del periodo la cual depende de la distancia d del centro hasta nuestro satélite en este caso y la velocidad de este

<h3 align="center">$ T = {2\pi d\over v} $,</h3>

Además de esto podemos obtener la velocidad 

<h3 align="center">$ F = {GMm\over d^2} $,</h3>

<h3 align="center">$ F = {p\over d} = {mv^2\over d} $,</h3>

<h3 align="center">$ {mv^2\over d} = {GMm\over d^2} $,</h3>

<h3 align="center">$ v = \sqrt{{GM\over d}} $,</h3>

Sustituyendo

<h3 align="center">$ T = {2\pi d\over \sqrt{{GM\over d}}} $,</h3>

Elevamos al cuadrado para quitarnos la raíz entonces nos queda como

<h3 align="center">$ T^2 = {4\pi^2 d^2\over {GM\over d}} $,</h3>

Despejamos la distancia d quedando como

<h3 align="center">$ d^3 = {GMT^2\over 4\pi^2} $,</h3>

Pero recordemos que la distancia d se esta considerando desde el centro de masa, en este caso el centro de la Tierra, por lo que tenemos que tomar en cuenta el radio de la Tierra y una altu5a h en la que se encuntre

<h3 align="center">$ (h+R)^3 = {GMT^2\over 4\pi^2} $,</h3>

<h3 align="center">$ h+R = \biggl( {GMT^2\over 4\pi^2} \biggr)^{1/3} $,</h3>

Por lo que nos queda como

<h3 align="center">$ h = \biggl( {GMT^2\over4\pi^2} \biggr)^{1/3} - R $,</h3>

(b) Escribe un programa que le pida al usuario que ingrese el valor deseado de $T$ y luego calcula e imprima la altitud correcta en metros.

In [51]:
# Programa que pida el valor esperado de T y calcule e imorima la altitud en metros
from math import pi

# Definimos las variable
G = (6.67)*(10**(-11))
M = (5.97)*(10**(24))
R = 6371000 #R ya en metros

# Pedimos el valor de T
T = float(input("Ingrese el valor de T:"))

# Calculamos h
h = ((((G*M)*(T**2))/(4*(pi**(2))))**(1/3))-R

# Impresión de la altura en metros
print("Dado el tiempo T dado como", T, "segundos, entonces la altura será h =", h, "metros.")

Ingrese el valor de T:87000
Dado el tiempo T dado como 87000.0 segundos, entonces la altura será h = 36051179.55892576 metros.


(c) Utiliza tu programa para calcular las altitudes de los satélites que orbitan la Tierra una vez al día (la llamada órbita *geoestacionaria*), una vez cada **90 minutos** y una vez cada **45 minutos**. ¿Qué concluyes de este último cálculo?

In [3]:
# Programa que para calcular las altitudes de los satélites que orbitan la Tierra una vez al día (la llamada órbita geoestacionaria), una vez cada 90 minutos y una vez cada 45 minutos.
from math import pi

#Definimos las variable
G = (6.67)*(10**(-11))
M = (5.97)*(10**(24))
R = 6371000 #R ya en metros

# Vamos a tomar ya valores conocidos de T, 90 minutos (x) y 45 minutos (y)
# Pasamos estos valores a segundos para obtener el resultado en metro

x = 90*60
y = 45*60

# La altura para el caso de 90 minutos será z y para los 45 minutos será w
# Calculamos las alturas con la ecuación ya conocida
z = ((((G*M)*(x**2))/(4*(pi**(2))))**(1/3))-R
w = ((((G*M)*(y**2))/(4*(pi**(2))))**(1/3))-R

#Impresión de las alturas en metros
print("La altitud del satélite que recorre la Tierra una vez cada 90 minutos es h(90) =", z, "metros, y la altitud del satélite que recorre la Tierra una vez cada 45 minutos es h(45) =", w, "metros.")

La altitud del satélite que recorre la Tierra una vez cada 90 minutos es h(90) = 279321.62537285965 metros, y la altitud del satélite que recorre la Tierra una vez cada 45 minutos es h(45) = -2181559.8978108233 metros.


Utilizando el código obtenemos que para 90 minutos 
    h(90)=279321.62537285965 metros
y para 45 minutos
    h(45)=-2181559.8978108233 metros

Se puede notar con este último valor que obtenemos un valor negativo, lo que a partir de la deducción de la fórmula, en específico

<h3 align="center">$ d= h+R = \biggl( {GMT^2\over 4\pi^2} \biggr)^{1/3} $,</h3>

Lo que signidfica que con este periodo T=45 minutos tendríamos que estría a una distancia d menor al rádio de la Tierra.

**3. Relatividad especial:** Una nave espacial viaja desde la Tierra en línea recta a una velocidad relativista $v$ a otro planeta a $x$ años luz de distancia.

 Escribe un programa que le pida al usuario el valor de $x$ y la velocidad $v$ como una fracción de la velocidad de la luz $c$, y que imprima el tiempo en años que tarda la nave espacial en llegar a su destino

a) en el marco de reposo de un observador en la Tierra y

In [4]:
# Programa que calcule el tiempo en años que tarda la nave en llegar al destino usando un valor de x y v

# Definimos las entradas
x = float(input("Ingresa la distancia entre la Tierra y el planeta destino en años luz x:"))
v = float(input("Ingresa la velocidad de la nave como fracción de la velocidad de la luz (valor entre 0 y 1) v"))

# Calculamos el tiempo
t = x/v

# Imprimimos el resultado
print("El tiempo en años que tarda la nave espacial en llegar a su destino en el marco de reposo de un observador en la Tierra ed de t = ", t, "años.")

Ingresa la distancia entre la Tierra y el planeta destino en años luz x:10
Ingresa la velocidad de la nave como fracción de la velocidad de la luz (valor entre 0 y 1) v0.8
El tiempo en años que tarda la nave espacial en llegar a su destino en el marco de reposo de un observador en la Tierra ed de t =  12.5 años.


b) como lo percibiría un pasajero a bordo de la nave.

Para el caso del pasajero a bordo de la nave tenemos que el percibirá que el permanece quieto mientras lo demás se mueve, aplicando relatividad especial podemos usar la transformación de Lorentz para la parte temporal 

<h3 align="center">$ t^\prime = {t\over\sqrt{1-v^2}}-{vx\over\sqrt{1-v^2}} $</h3>

Donde la t primada será el tiempo a calcular en años luz en el programa, v la velocidad en fracción de la velocidad de la luz c, t el tiempo desde el sistema de la Tierra y x la distancia entre la Tierra y el planeta.

In [5]:
# Programa que calcule el tiempo en años que tarda la nave en llegar al destino usando un valor de x y v, percibido por el astronauta
import math

# Definimos las entradas
x = float(input("Ingresa la distancia entre la Tierra y el planeta destino en años luz x:"))
v = float(input("Ingresa la velocidad de la nave como fracción de la velocidad de la luz (valor entre 0 y 1) v"))

# Calculamos el tiempo desde la tierra
t1 = x/v

# Usamos la transformación de Lorentz

t2 = ((t1/(math.sqrt(1-(v**2))))-((v*x)/(math.sqrt(1-(v**2)))))

# Imprimimos el resultado
print("El tiempo en años que tarda la nave espacial en llegar a su destino en el marco de del astronauta en la nave es t = ", t2, "años.")

Ingresa la distancia entre la Tierra y el planeta destino en años luz x:10
Ingresa la velocidad de la nave como fracción de la velocidad de la luz (valor entre 0 y 1) v0.99
El tiempo en años que tarda la nave espacial en llegar a su destino en el marco de del satronauta en la nave es t =  1.4249228262288653 años.


Usa tu programa para calcular las respuestas para un planeta a 10 años luz de distancia con $v = 0.99c$

In [6]:
# Programa que calcule el tiempo en años que tarda la nave en llegar al destino a 10 años luz de distancia con  𝑣=0.99𝑐

# Definimos las entradas
x = 10
v = 0.99

# Calculamos el tiempo
t = x/v

# Imprimimos el resultado
print("El tiempo en años que tarda la nave espacial en llegar a su destino en el marco de reposo de un observador en la Tierra con velocidad de 0.99c y a una distancia de 10 años luz es de t = ", t, "años.")

El tiempo en años que tarda la nave espacial en llegar a su destino en el marco de reposo de un observador en la Tierra con velocidad de 0.99c y a una distancia de 10 años luz es de t =  10.1010101010101 años.


In [7]:
# Programa que calcule el tiempo en años que tarda la nave en llegar al destino a 10 años luz de distancia con  𝑣=0.99𝑐, percibido por el astronauta
import math

# Definimos las entradas
x = 10
v = 0.99

# Calculamos el tiempo desde la tierra
t1 = x/v

# Usamos la transformación de Lorentz

t2 = ((t1/(math.sqrt(1-(v**2))))-((v*x)/(math.sqrt(1-(v**2)))))

# Imprimimos el resultado
print("El tiempo en años que tarda la nave espacial en llegar a su destino en el marco de del satronauta en la nave con velocidad de 0.99c y a una distancia de 10 años luz es t = ", t2, "años.")

El tiempo en años que tarda la nave espacial en llegar a su destino en el marco de del satronauta en la nave con velocidad de 0.99c y a una distancia de 10 años luz es t =  1.4249228262288653 años.


**4. Órbitas planetarias:** En el espacio, la órbita de un cuerpo alrededor de otro (como un planeta alrededor del Sol), no necesariamente es circular. En general, toma la forma de una elipse, con el cuerpo a veces más cerca y otras más lejos. Si tenemos la distancia $\ell_1$ de máxima aproximación de un planeta al Sol (su *perihelio*), y su velocidad lineal $v_1$ en el perihelio, entonces cualquier otra propiedad de la órbita se puede calcular a partir de estas dos cantidades de la siguiente manera:

   
**a)** La segunda ley de Kepler nos dice que la distancia $\ell_2$ y la velocidad $v_2$ del planeta en su punto más distante, o *afelio*, satisfacen que $\ell_2 v_2 = \ell_1 v_1$.  Al mismo tiempo, la energía total, cinética más la gravitatoria, de un planeta con velocidad $v$ y distancia $r$ del Sol está dada por:

<h3 align="center">$E = \frac{1}{2} m v^2 - G {mM\over r}$,</h3>
        
donde $m$ es la masa del planeta, $M=1.9891\times10^{30}\,$kg es la masa del Sol y $G=6.6738\times10^{-11}\,\mathrm{m^3\,kg^{-1}\,s^{-2}}$ es la constante gravitatoria de Newton. Dado que la energía debe conservarse, demuestre que $v_2$ es la raíz más pequeña de la ecuación cuadrática:

<h3 align="center">$v_2^2 - {2GM\over v_1\ell_1} v_2 - \biggl[ v_1^2 - {2GM\over\ell_1} \biggr] = 0$.</h3>

Una vez que tenemos $v_2$ podemos calcular $\ell_2$  usando la relación $\ell_2 = \ell_1 v_1/v_2$.

**Demostración**

Partimos de la ecuación de conservación de la energía, ya que se debe de conservar sin importar que nos encontremos en el punto 1 o en el punto 2 tendremos que

<h3 align="center">$ E_{1} = E_{2} $</h3>

<h3 align="center">$ \frac{1}{2} m v_{1}^2 - G {mM\over \ell_{1}} = \frac{1}{2} m v_{2}^2 - G {mM\over \ell_{2}} $</h3>

De aquí multiplicamos por 2 y dividimos entre m toda la ecuación, ya que es la misma masa y el 2 es una constante

<h3 align="center">$ v_{1}^2 - 2G {mM\over m\ell_{1}} = v_{2}^2 - 2G {mM\over m\ell_{2}} $</h3>

Cancelando las m, ahora sustituimos $\ell_2$ usando que

<h3 align="center">$ \ell_2 = {\ell_1 v_{1}\over v_{2}} $</h3>

tal que

<h3 align="center">$ v_{1}^2 - 2G {M\over \ell_{1}} = v_{2}^2 - 2G {M v_{2}\over \ell_1 v_{1}} $</h3>

Igualando a 0 tendremos 

<h3 align="center">$v_2^2 - {2GM\over v_1\ell_1} v_2 -  v_1^2 +{2GM\over\ell_1} = 0$.</h3>

<h3 align="center">$v_2^2 - {2GM\over v_1\ell_1} v_2 - \biggl[ v_1^2 - {2GM\over\ell_1} \biggr] = 0$.</h3>

De donde podemos observar que lo obtenido es una ecuación de segundo grado de la cual podemos obtener el resultado de $v_2$ tal que 

<h3 align="center">$v_2 = {{{2GM\over v_1\ell_1} + \sqrt{\biggl( {2GM\over v_1\ell_1} \biggr)^2 + 4v_1^2 - {8GM\over\ell_1}} }\over 2} $.</h3>

De donde solo necesitamos valores de $v_1$ y de $\ell_1$ para obtener el valor de $v_2$, y a partir de este el de $\ell_2$.

**b)** Dados los valores de $v_1$, $\ell_1$, y $\ell_2$; otros parámetros de la órbita se obtienen mediante fórmulas simples que pueden derivarse de las leyes de Kepler y del hecho de que la órbita es una elipse:

<h4 align="center">Semieje mayor: $\; \; \; a = \frac{1}{2}(\ell_1+\ell_2)$,</h4>

<h4 align="center">Semieje menor: $\; \; \; b = \sqrt{\ell_1\ell_2}$,</h4>

<h4 align="center">Período orbital: $\; \; \; T = {2\pi ab\over\ell_1 v_1}$,</h4>

<h4 align="center">Excentricidad orbital: $\; \; \; e = {\ell_2-\ell_1\over\ell_2+\ell_1}$.</h4>

Escribe un programa que le pida al usuario que ingrese la distancia al Sol y la velocidad en el perihelio; para que calcule e imprima las cantidades:

$\ell_2$, $v_2$, $T$ y $e$.

In [164]:
# Programa que calcule v2, l2, T y e a partir de v1 y l1
from math import pi

# Definimos las entradas
l1 = float(input("Ingresa la distancia la distancia al Sol en el perihelio en metros: "))
v1 = float(input("Ingresa la velocidad en el perihelio en m/s: "))
G = 6.6738*(10**(-11))
M = 1.9891*(10**(30))

#Calculamos v2
v2 = ((2*G*M)/(v1*l1)) - v1

# Usando v2 calculamos l2
l2 = (l1*v1)/v2

# Calculamos a y b
a = (1/2)*(l1+l2)
b = ((l1*l2))**(1/2)

# Calculamos T
T = (2*pi*a*b)/(l1*v1)

# Calculamos e
e = (l2-l1)/(l2+l1)

# Imprimimos los reusltados
print("Velocidad del planeta en el afelio, v2 =", v2, "m/s")
print("Distancia del planeta en el afelio, l2 =", l2, "m")
print("Periodo orbital del planeta, T =", T, "s")
print("Excentricidad orbital del planeta, e =", e)


Ingresa la distancia la distancia al Sol en el perihelio en metros: 148000000000
Ingresa la velocidad en el perihelio en m/s: 30
Velocidad del planeta en el afelio, v2 = 59796616.75675675 m/s
Distancia del planeta en el afelio, l2 = 74251.69250061794 m
Periodo orbital del planeta, T = 10977738.272935234 s
Excentricidad orbital del planeta, e = -0.9999989965992533


**c)** Prueba tu programa haciendo que calcule las propiedades de las órbitas de la Tierra (para las cuales $\ell_1=1.4710\times10^{11}\,$m y $v_1=3.0287\times10^4\,\mathrm{m\,s^{-1}}$)  y del cometa Halley ($\ell_1=8.7830\times10^{10}\,$m y $v_1=5.4529\times10^4\,\mathrm{m\,s^{-1}}$). Entre otras cosas, deberías encontrar que el período orbital de la Tierra es de un año y el del cometa Halley es de unos 76 años.

In [169]:
# Programa que calcule v2, l2, T y e a partir de v1 y l1 para la Tierra y el cometa Halley
from math import pi

# Definimos las variables
G = 6.6738*(10**(-11))
M = 1.9891*(10**(30))

##############################################################################

# Para la Tierra
lt1 = 1.4710*(10**11)
vt1 = 30287

#Calculamos v2
vt2 = ((2*G*M)/(vt1*lt1)) - vt1

# Usando v2 calculamos l2
lt2 = (lt1*vt1)/vt2

# Calculamos a y b
at = (1/2)*(lt1+lt2)
bt = ((lt1*lt2))**(1/2)

# Calculamos T
Tt = (2*pi*at*bt)/(lt1*vt1)

# Calculamos e
et = (lt2-lt1)/(lt2+lt1)

##############################################################################

# Para el cometa Halley
l1 = 8.7830*(10**10)
v1 = 5.4529*(10**4)

#Calculamos v2
v2 = ((2*G*M)/(v1*l1)) - v1

# Usando v2 calculamos l2
l2 = (l1*v1)/v2

# Calculamos a y b
a = (1/2)*(l1+l2)
b = ((l1*l2))**(1/2)

# Calculamos T
T = (2*pi*a*b)/(l1*v1)

# Calculamos e
e = (l2-l1)/(l2+l1)

##############################################################################

# Imprimimos los reusltados
print("Para la Tierra:") 
print("La velocidad en el afelio, v2 =", vt2, "m/s")
print("La distancia de la Tierra en el afelio, l2 =", lt2, "m")
print("El periodo orbital de la Tierra, T =", Tt, "s")
print("La excentricidad orbital de la Tierra, e =", et)
print("Para el cometa Halley:")
print("La velocidad en el afelio, v2 =", v2, "m/s")
print("La distancia del cometa en el afelio, l2 =", l2, "m")
print("El periodo orbital del cometa, T =", T, "s")
print("La excentricidad orbital del cometa, e =", e)

Para la Tierra:
La velocidad en el afelio, v2 = 29305.3991772613 m/s
La distancia de la Tierra en el afelio, l2 = 152027197208.6598 m
El periodo orbital de la Tierra, T = 31543060.207886893 s
La excentricidad orbital de la Tierra, e = 0.016471913134741688
Para el cometa Halley:
La velocidad en el afelio, v2 = 906.6806969191457 m/s
La distancia del cometa en el afelio, l2 = 5282214660876.463 m
El periodo orbital del cometa, T = 2399312511.8452024 s
La excentricidad orbital del cometa, e = 0.9672889126454062


**5. La fórmula semiempírica de la masa (FSM)**

En física nuclear, la **fórmula de Weizsäcker** (conocida también como fórmula semiempírica) sirve para evaluar la masa y otras propiedades de un núcleo atómico; y está basada parcialmente en mediciones empíricas. En particular la fórmula se usa para calcular la **energía de enlace nuclear aproximada** $B$, de un núcleo atómico con número atómico $Z$ y número de masa $A$:

<h3 align="center">$B = a_1 A - a_2 A^{2/3} - a_3 {Z^2\over A^{1/3}} - a_4 {(A - 2Z)^2\over A} + {a_5\over A^{1/2}}$,</h3>

donde, en unidades de millones de electrón-volts, las constantes son $a_1=15.8$, $a_2=18.3$, $a_3=0.714$, $a_4=23.2$ y          
<h6 align="center">
$a_5 =
  \left \{
    \begin{aligned}
      0 & \; \text{si} \ A \; \text{es impar,} \\
      12.0 & \; \text{si} \ A \; \text{y} \; Z \; \text{son pares (ambos),} \\
      -12.0 & \; \text{si} \ A \; \text{es par y} \; Z \; \text{impar}
    \end{aligned}
  \right.
$
</h6>

**a)** Escribe un programa que tome como entrada los valores de $A$ y $Z$, e imprima la energía de enlace $B$ para el átomo correspondiente. Usa tu programa para encontrar la energía de enlace de un átomo con $A=58$ y $Z=28$ (Hint: La respuesta correcta es alrededor de los $490\,$MeV).

In [1]:
# Programa que tome como entrada los valores de  𝐴 y  𝑍, e imprima la energía de enlace  𝐵

# Pedimos los valores de A y Z
A = float(input("Ingresa el número de masa A: "))
Z = float(input("Ingresa el número atómico Z: "))

# Definimos las cosntantes a fijas
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2

#Ahora definimos el valor de a5
if A%2==0 and Z%2==0:
    a5 = 12
elif A%2==0 and Z%2==1:
    a5 = -12
elif A%2==1:
    a5 = 0

# Calculamos B
B = (a1*A)-(a2*(A**(2/3)))-(a3*((Z**2)/(A**(1/3))))-(a4*(((A-(2*Z))**2)/A))+(a5/(A**(1/2)))

print("La energía de enlace nuclear aproximada B =", B, "MeV.")

Ingresa el número de masa A: 58
Ingresa el número atómico Z: 28
La energía de enlace nuclear aproximada B = 497.5620206224374 MeV.


**b)** Modifica el programa del inciso anterior, para escribir una segunda versión que imprima no la energía de enlace total $B$, sino la energía de unión por nucleón, que es $B/A$.

In [12]:
# Programa para escribir una segunda versión que imprima la energía de unión por nucleón, que es 𝐵/𝐴

# Pedimos los valores de A y Z
A = float(input("Ingresa el número de masa A: "))
Z = float(input("Ingresa el número atómico Z: "))

# Definimos las cosntantes a fijas
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2

#Ahora definimos el valor de a5
if A%2==0 and Z%2==0:
    a5 = 12
elif A%2==0 and Z%2==1:
    a5 = -12
elif A%2==1:
    a5 = 0

# Calculamos B
B = (a1*A)-(a2*(A**(2/3)))-(a3*((Z**2)/(A**(1/3))))-(a4*(((A-(2*Z))**2)/A))+(a5/(A**(1/2)))

# Ahora solo dividimos entre A el resultado anterior 
E = B/A

print("La energía de de unión por nucleón, que es 𝐵/𝐴 =", E, "MeV.")

Ingresa el número de masa A: 58
Ingresa el número atómico Z: 28
La energía de de unión por nucleón, que es 𝐵/𝐴 = 8.578655527973059 MeV.


**c)** Escribe una tercera versión del programa para que tome como entrada solo un valor del número atómico $Z$ y luego pase por todos los valores de $A$ desde $A=Z$ hasta $A=3Z$, para encontrar el que tiene la mayor energía de enlace por nucleón . Este es el núcleo más estable con el número atómico dado. Haz que tu programa imprima el valor de $A$ para este núcleo más estable y el valor de la energía de enlace por nucleón.

In [18]:
# Programa para encontrar a partir de Z el valor de A con mayor energía de enlace por nucleón B/A
from numpy import array


# Pedimos los valores de Z
Z = int(input("Ingresa el número atómico Z: "))

# Definimos las cosntantes a fijas
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2

# Creamos listas vacías para poner los valores de E y A generaods del ciclo for
T = []
C = []

# Usamos el ciclo for además de las mismas especificaciones para obtener el valor de E
for n in range(1, 4):
    A = n*Z
    if A%2==0 and Z%2==0:
        a5 = 12
    elif A%2==0 and Z%2==1:
        a = -12
    elif A%2==1:
        a5 = 0
    B = (a1*A)-(a2*(A**(2/3)))-(a3*((Z**2)/(A**(1/3))))-(a4*(((A-(2*Z))**2)/A))+(a5/(A**(1/2)))
    E = B/A
    
# Agregamos los valores de E y A a las listas
    T.append(E)
    C.append(A)
    
# Buscamos el valor máximo de E así como su posición para obtener el valor de la lista de A que coincida con esa posición
h = max(T)
x = T.index(h)
y = C[x]

# Imprimimos el resultado
print("La energía de de unión por nucleón máxima, que es 𝐵/𝐴 =", h, "MeV, con A =", y)

Ingresa el número atómico Z: 28
La energía de de unión por nucleón máxima, que es 𝐵/𝐴 = 8.432655535147047 MeV, con A = 56


**d)** Finalmente, escribe una cuarta versión del programa que, en lugar de tomar $Z$ como entrada, se ejecute a través de todos los valores de $Z$ de 1 a 100 e imprima el valor más estable de $A$ para cada uno. ¿A qué valor de $Z$ se produce la energía de enlace máxima por nucleón? (La respuesta correcta, en la vida real, es $Z=28$, que corresponde al Níquel).

In [163]:
# Programa que calcula  el valor más estable de  𝐴 para cada uno

import math
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2
biggerBn = 0.0
biggerZ = 0.0
for m in list(range(1,101)):
    Z = m
    bigA = 0.0
    bigBn = 0.0
    for n in list(range(1,4)):
        A = Z*n
        if A%2 == 1:
            a5 = 0
        elif Z%2 == 1:
            a5 = -12
        else:
            a5 = 12
        B = a1*A - a2*(A**(2/3)) - a3*((Z**2)/(A**(1/3))) - a4*(((A - 2*Z)**2)/A) + a5*(A**(-1/2))
        Bn = B/A
        #print(Bn)
        if Bn > bigBn:
            bigBn = Bn
            bigA = A
            bigZ = Z
        if bigBn > biggerBn:
            biggerBn = bigBn
            biggerZ = bigZ
    print("Los valores de A con la energía mas estable para cada grupo son", bigA)
print("Energía de enlace máxima por nucleón", biggerZ)
   

Los valores de A con la energía mas estable para cada grupo son 3
Los valores de A con la energía mas estable para cada grupo son 4
Los valores de A con la energía mas estable para cada grupo son 6
Los valores de A con la energía mas estable para cada grupo son 8
Los valores de A con la energía mas estable para cada grupo son 10
Los valores de A con la energía mas estable para cada grupo son 12
Los valores de A con la energía mas estable para cada grupo son 14
Los valores de A con la energía mas estable para cada grupo son 16
Los valores de A con la energía mas estable para cada grupo son 18
Los valores de A con la energía mas estable para cada grupo son 20
Los valores de A con la energía mas estable para cada grupo son 22
Los valores de A con la energía mas estable para cada grupo son 24
Los valores de A con la energía mas estable para cada grupo son 26
Los valores de A con la energía mas estable para cada grupo son 28
Los valores de A con la energía mas estable para cada grupo son 30